client/src/AppBundle/Service/Audit/AwsAuditLogHandler.php

```diff
@@ -0,0 +1,174 @@
+            $this->initialize();
+        }
+
+        // send items, retry once with a fresh sequence token
+        try {
+            $this->send($entry);
+        } catch (CloudWatchLogsException $e) {
+            $this->determineSequenceToken($refresh = true);
+            $this->send($entry);
+        }
+    }
+
+    /**
+     * @param array $entry
+     * @return array
+     */
+    private function formatEntry(array $entry): array
+    {
+        return [
+            [
+                'message' => $entry['formatted'],
+                'timestamp' => $entry['datetime']->format('U.u') * 1000
+            ]
+        ];
+    }
+
+
+    private function initialize(): void
+    {
+        $this->existingStreams = $this->fetchExistingStreams();
+        $existingStreamsNames = $this->extractExistingStreamNames();
+
+        if (!in_array($this->stream, $existingStreamsNames, true)) {
+            $this->createLogStream();
+        } else {
+            $this->determineSequenceToken();
+        }
+
+        $this->initialized = true;
+    }
+
+    /**
+     * @return array
+     */
+    private function fetchExistingStreams(): array
+    {
+        return $this
+            ->client
+            ->describeLogStreams(
+                [
+                    'logGroupName' => $this->group,
+                    'logStreamNamePrefix' => $this->stream,
+                ]
+            )->get('logStreams');
+    }
+
+    /**
+     * @return array
+     */
+    private function extractExistingStreamNames(): array
+    {
+        return array_map(
+            function ($stream) {
+                return $stream['logStreamName'];
+            },
+            $this->existingStreams
+        );
+    }
+
+    private function createLogStream(): void
+    {
+        $this
+            ->client
+            ->createLogStream(
+                [
+                    'logGroupName' => $this->group,
+                    'logStreamName' => $this->stream
+                ]
+            );
+    }
+
+    /**
+     * @param bool $refresh
+     */
+    private function determineSequenceToken(bool $refresh = false): void
+    {
+        if ($refresh) {
+            $this->existingStreams = $this->fetchExistingStreams();
+        }
+
+        foreach ($this->existingStreams as $stream) {
+            if ($stream['logStreamName'] === $this->stream && isset($stream['uploadSequenceToken'])) {
+                $this->sequenceToken = $stream['uploadSequenceToken'];
+                break;
+            }
+        }
+    }
+
+    /**
+     * @param array $entry
+     */
+    private function send(array $entry): void
+    {
+        $data = [
+            'logGroupName' => $this->group,
+            'logStreamName' => $this->stream,
+            'logEvents' => $entry
+        ];
+
+        if (!empty($this->sequenceToken)) {
+            $data['sequenceToken'] = $this->sequenceToken;
+        }
+
+        $response = $this->client->putLogEvents($data);
```

Comment:
How do we handle a failed response due to connectivity?

Reply:
The AWS SDK implements automatic retries with exponential backoffs for increasing delay periods between retries. This particular client instance is using the default number of retries which is 3.

Reply:
Cool thought it might be under the hood, but wasn't sure.